hal/armv7m/mpu.c

```diff
@@ -0,0 +1,333 @@
+}
+
+
+static void mpu_attrPrint(u32 attr, int enabled)
+{
+	static const char *strAP[8] = { "na/na", "rw/na", "rw/ro", "rw/rw", "reserved", "ro/ro", "ro/ro" };
+	static const char *strYesNo[2] = { "n", "y" };
+
+	/* enabled */
+	lib_printf("%-4s", strYesNo[enabled != 0]);
+
+	/* eXec Never */
+	lib_printf("%-4s", strYesNo[(attr >> 12) & 1ul]);
+
+	/* Access Permission */
+	lib_printf("%-12s", strAP[(attr >> 8) & 0x7ul]);
+
+	/* TEX */
+	lib_printf("%c", '0' + ((attr >> 5) & 1ul));
+	lib_printf("%c", '0' + ((attr >> 4) & 1ul));
+	lib_printf("%c  ", '0' + ((attr >> 3) & 1ul));
+
+	/* Shareable */
+	lib_printf("%-3s", strYesNo[(attr >> 2) & 1ul]);
+
+	/* Cacheeable */
+	lib_printf("%-3s", strYesNo[(attr >> 1) & 1ul]);
+
+	/* Bufferable */
+	lib_printf("%s\n", strYesNo[attr & 1ul]);
+}
+
+
+/* Describe content of rbar & rasr registers */
+static void mpu_regionPrint(const char *name, u32 rbar, u32 rasr)
+{
+	unsigned int bit;
+	u32 srSize, srBase = rbar & (0x7fffffful << 5);
+	u32 attr = (rasr >> 16) & 0x173f;
+	u8 sizeBit = ((rasr >> 1) & 0x1fu) + 1;
+	u8 srdMask = rasr >> 8;
+	u8 region = rbar & 0xfu;
+
+	if (srdMask != 0) {
+		srSize = 1u << (sizeBit - 3);
+
+		for (bit = 0; bit < 8; bit++) {
+			if ((srdMask & (1u << bit)) == 0) {
+				lib_printf("%-10s %d%-5s %d%-8s 0x%08x%4s 0x%08x%4s",
+					name, region, "", bit, "", srBase, "", srBase + srSize - 1, "");
+				mpu_attrPrint(attr, (rasr & 1));
+			}
+
+			srBase += srSize;
+		}
+	}
+	else {
+		lib_printf("%-10s %d%-5s %-9s 0x%08x%4s 0x%08x%4s",
+			name, region, "", "all", srBase, "", srBase + (1ul << sizeBit), "");
+		mpu_attrPrint(attr, (rasr & 1));
+	}
+}
+
+
+/* Mark region valid, but disabled */
+static void mpu_regionDisable(u8 idx)
+{
+	if (idx < mpu_common.regMax) {
+		mpu_common.hal.mpu.region[idx].rbar =
+			((mpu_common.hal.mpu.region[idx].rbar | (1ul << 4)) & ~0xful) | (idx & 0xful);
+		mpu_common.hal.mpu.region[idx].rasr &= ~1;
+	}
+}
+
+
+/* Setup single MPU region entry in a local MPU context */
+static int mpu_regionSet(u8 idx, u32 addr, u8 sizeBit, u32 attr, u8 enable, u8 srdMask)
+{
+	u8 tex = 0; /* FIXME: TEX is no used currently, defaults to 0 */
+
+	u8 ap = 1; /* unprivileged no access */
+
+	if (sizeBit < 5 || idx >= mpu_common.regMax)
+		return -1;
+
+	if (attr & mAttrRead)
+		ap = 2; /* unprivileged read only access */
+
+	if (attr & mAttrWrite)
+		ap = 3; /* unprivileged read and write access */
+
+	/* privileged access have read-write access */
+
+	/* clang-format off */
+
+	mpu_common.hal.mpu.region[idx].rbar =
+		(addr & (0x7fffffful << 5))
+		| (1ul << 4)
+		| (idx & 0xful);
+
+	mpu_common.hal.mpu.region[idx].rasr =
+		((((attr & mAttrExec) == 0) & 1ul) << 28)
+		| ((ap & 0x7ul) << 24)
+		| ((tex & 0x7ul) << 19)
+		| ((((attr & mAttrShareable) != 0) & 1ul) << 18)
+		| ((((attr & mAttrCacheable) != 0) & 1ul) << 17)
+		| ((((attr & mAttrBufferable) != 0) & 1ul) << 16)
+		| ((srdMask & 0xfful) << 8)
+		| (((sizeBit - 1) & 0x1ful ) << 1)
+		| (enable != 0);
```

Comment:
Our coding style requires binary operands to be last character on broken lines. That's why you had to disable clang-format